examples/virtual_methods/base_button/imp.rs

```diff
@@ -0,0 +1,115 @@
+use super::BaseButtonExt;
+use gtk::{
+    gio,
+    glib::{self, translate::from_glib_borrow, Error},
+    prelude::*,
+    subclass::prelude::*,
+};
+use std::{future::Future, pin::Pin};
+
+pub type BaseButtonInstance = <BaseButton as super::ObjectSubclass>::Instance;
+pub type PinnedFuture = Pin<Box<dyn Future<Output = Result<(), Error>> + 'static>>;
```

Comment:
I'm surprised this doesn't cause a warning. It's not really an FFI-safe type, is it?

Reply:
Hm, apparently this causes a warning on nightly but not on stable, interesting.

Reply:
Ah, I suppose only nightly understands the problem here, since this _is_ FFI save for `Sized` types (but since it's a `dyn T` we'll need pointers, as you've said)

See: 
https://doc.rust-lang.org/std/boxed/index.html#memory-layout

Reply:
Hm, `*mut dyn Future<Output = T>>` isn't FFI save either though

Reply:
`*mut Box<dyn Future<_>>` is what you need (double boxing)

Reply:
Oh, makes sense, thanks :)

Reply:
FWIW, as this is all called only from Rust, you can make all the vfuncs also without the `extern "C"` and use the Rust ABI and proper Rust types everywhere.

Only if you want it to be callable from C that would be needed.

Reply:
Oh, I had figured I'd need that somehow if GObject interacted with it

Reply:
For GObject that's just some struct field in either way. Only if you need other languages than Rust to interpret it, then you need it to be `extern "C"`.